trace_decoder/src/processed_block_trace.rs

```diff
@@ -266,131 +272,169 @@ impl<F: CodeHashResolveFunc> CodeHashResolving<F> {
-                .map(|(k, v)| (Nibbles::from_h256_be(k), rlp::encode(&v).to_vec()))
-                .collect();
+                let storage_read_keys = trace
+                    .storage_read
+                    .clone()
+                    .into_iter()
+                    .flat_map(|reads| reads.into_iter());
 
-            nodes_used_by_txn
-                .storage_writes
-                .push((hashed_addr, storage_writes_vec));
+                let storage_write_keys = storage_writes.keys();
+                let storage_access_keys = storage_read_keys.chain(storage_write_keys.copied());
 
-            nodes_used_by_txn.state_accesses.push(hashed_addr);
+                if let Some(storage) = nodes_used_by_txn.storage_accesses.get_mut(&hashed_addr) {
+                    storage.extend(
+                        &storage_access_keys
+                            .map(|k| Nibbles::from_h256_be(hash(&k.0)))
+                            .collect::<Vec<Nibbles>>(),
+                    )
+                } else {
+                    nodes_used_by_txn.storage_accesses.insert(
+                        hashed_addr,
+                        storage_access_keys
+                            .map(|k| Nibbles::from_h256_be(hash(&k.0)))
+                            .collect::<Vec<Nibbles>>(),
+                    );
+                };
 
-            if let Some(c_usage) = trace.code_usage {
-                match c_usage {
-                    ContractCodeUsage::Read(c_hash) => {
-                        contract_code_accessed
-                            .entry(c_hash)
-                            .or_insert_with(|| code_hash_resolver.resolve(&c_hash));
+                let storage_trie_change = !storage_writes.is_empty();
+                let code_change = trace.code_usage.is_some();
+                let state_write_occurred = trace.balance.is_some()
+                    || trace.nonce.is_some()
+                    || storage_trie_change
+                    || code_change;
+
+                if state_write_occurred {
+                    if let Some(state_trie_writes) =
+                        nodes_used_by_txn.state_writes.get_mut(&hashed_addr)
+                    {
+                        // The entry already exists, so we update only the relevant fields.
+                        if trace.balance.is_some() {
+                            state_trie_writes.balance = trace.balance;
+                        }
+                        if trace.nonce.is_some() {
+                            state_trie_writes.nonce = trace.nonce;
+                        }
+                        if storage_trie_change {
+                            state_trie_writes.storage_trie_change = storage_trie_change;
+                        }
+                        if code_change {
+                            state_trie_writes.code_hash =
+                                trace.code_usage.as_ref().map(|usage| usage.get_code_hash());
+                        }
```

Comment:
Minor nit: I wonder if we can create a helper function here like `update_val_if_some`? Or maybe a version of this function called `update_val_if_true`?

```rust
pub(crate) fn update_val_if_some<T>(target: &mut T, opt: Option<T>) {
    if let Some(new_val) = opt {
        *target = new_val;
    }
}
```

Reply:
Yeah that'd be cleaner. I did this on the fly while debugging but it's kinda ugly currently